File: app/src/main/java/fr/xgouchet/chronorg/data/repositories/BaseRepository.java
package fr.xgouchet.chronorg.data.repositories;

import android.content.ContentResolver;
import android.content.Context;
import android.support.annotation.NonNull;

import fr.xgouchet.chronorg.data.ioproviders.IOProvider;
import rx.Observable;
import rx.Subscriber;

/**
 * @author Xavier Gouchet
 */
public class BaseRepository<T> {

    @NonNull /*package*/ final Context context;

    @NonNull /*package*/ final IOProvider<T> provider;

    public BaseRepository(@NonNull Context context, @NonNull IOProvider<T> provider) {
        this.context = context;
        this.provider = provider;
    }

    public Observable<T> getAll() {

        return Observable.create(new Observable.OnSubscribe<T>() {
            @Override public void call(final Subscriber<? super T> subscriber) {
                try {
                    ContentResolver contentResolver = context.getContentResolver();
                    provider.provideQuerier().queryAll(contentResolver, new SubscriberWrapperAction<>(subscriber));
                    subscriber.onCompleted();
                } catch (Exception err) {
                    subscriber.onError(err);
                }
            }
        });
    }

    public Observable<T> get(final int id) {
        return Observable.create(new Observable.OnSubscribe<T>() {
            @Override public void call(final Subscriber<? super T> subscriber) {
                try {
                    ContentResolver contentResolver = context.getContentResolver();
                    provider.provideQuerier().query(contentResolver, new SubscriberWrapperAction<>(subscriber), id);
                    subscriber.onCompleted();
                } catch (Exception err) {
                    subscriber.onError(err);
                }
            }
        });
    }

    public Observable<Void> save(@NonNull final T item) {

        return Observable.create(new Observable.OnSubscribe<Void>() {
            @Override public void call(Subscriber<? super Void> subscriber) {
                try {
                    ContentResolver contentResolver = context.getContentResolver();
                    boolean success = provider.provideQuerier().save(contentResolver, item);
                    if (success) {
                        subscriber.onCompleted();
                    } else {
                        subscriber.onError(new RuntimeException("Unable to save item !"));
                    }
                } catch (Exception e) {
                    subscriber.onError(e);
                }

            }
        });
    }

    public Observable<Void> delete(final @NonNull T item) {
        return Observable.create(new Observable.OnSubscribe<Void>() {
            @Override public void call(Subscriber<? super Void> subscriber) {
                try {
                    ContentResolver contentResolver = context.getContentResolver();
                    boolean success = provider.provideQuerier().delete(contentResolver, item);
                    if (success) {
                        subscriber.onCompleted();
                    } else {
                        subscriber.onError(new RuntimeException("Unable to delete item !"));
                    }
                } catch (Exception e) {
                    subscriber.onError(e);
                }

            }
        });
    }
}

File: khronorg/src/main/java/fr/xgouchet/khronorg/data/repositories/BaseRepository.kt
package fr.xgouchet.khronorg.data.repositories

import android.content.Context
import fr.xgouchet.khronorg.data.ioproviders.IOProvider
import fr.xgouchet.khronorg.data.query.QueryAlteration
import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe
import io.reactivex.functions.Consumer
import io.reactivex.subjects.BehaviorSubject

/**
 * @author Xavier F. Gouchet
 */
class BaseRepository<T>(val context: Context, val provider: IOProvider<T>) : Repository<T> {

    private val currentItemSource = BehaviorSubject.create<T>()

    override fun current(): Observable<T> {
        return currentItemSource
    }

    override fun setCurrent(item: T) {
        currentItemSource.onNext(item)
    }

    override fun getCurrent(): T {
        return currentItemSource.value
    }

    override fun getAll(): Observable<T> {
        val source = ObservableOnSubscribe<T> {
            emitter ->
            try {
                val contentResolver = context.contentResolver
                provider.querier.queryAll(contentResolver, Consumer { t -> emitter.onNext(t) })
                emitter.onComplete()
            } catch (e: Exception) {
                emitter.onError(e)
            }
        }

        return Observable.create(source)
    }


    override fun getWhere(alter: QueryAlteration): Observable<T> {
        val source = ObservableOnSubscribe<T> {
            emitter ->
            try {
                val contentResolver = context.contentResolver
                provider.querier.queryWhere(contentResolver, alter, Consumer { t -> emitter.onNext(t) })
                emitter.onComplete()
            } catch (e: Exception) {
                emitter.onError(e)
            }
        }

        return Observable.create(source)
    }

    override fun save(item: T): Observable<Any> {
        val project = item
        val source = ObservableOnSubscribe<Any> {
            emitter ->
            try {
                val contentResolver = context.contentResolver
                if (provider.querier.save(contentResolver, project)) {
                    emitter.onComplete()
                } else {
                    emitter.onError(RuntimeException("Can't save… WTF‽"))
                }
            } catch (e: Exception) {
                emitter.onError(e)
            }
        }

        return Observable.create(source)
    }
}